1 The Physical/Guess Point Level 1 (Stateless)/2 - Arrange-Act-Assert Backwards/2.1 - Password Validator/src/index.spec.ts

```diff
@@ -1,5 +1,84 @@
-
-describe('password validator', () => {
-
-})
-
+import { PasswordValidator } from "./password-validator";
+import { Errors } from "./errors";
+
+describe('password validator', () => {
+
+    let passwordValidator: PasswordValidator;
+    beforeEach(() => {
+        passwordValidator = new PasswordValidator()
+    })
+
+    it.each([
+        ['A non valid'],
+        ['Other non valid'],
+        ['Anothernonvalid'],
+    ])('given the password {%s} when call the password validator then it should return an invalid response with NO_DIGITS error', (invalidPassword: string) => {
+        const result = passwordValidator.validate(invalidPassword);
+        expect(result.status).toBeFalsy();
+        expect(result.errors).toEqual([Errors.NO_DIGITS]);
+    });
+
+    it.each([
+        ['1 non valid'],
+        ['2 non valid'],
+        ['3 non valid'],
+        ['32 non valid'],
+    ])('given the password {%s} when call the password validator then it should return an invalid response with NO_UPPER error', (invalidPassword: string) => {
+        const result = passwordValidator.validate(invalidPassword);
+        expect(result.status).toBeFalsy();
+        expect(result.errors).toEqual([Errors.NO_UPPER]);
+    });
+
+    it.each([
+        ['A1fe'],
+        ['B2fe'],
+        ['C3fe'],
+        ['C3fE'],
+    ])('given a short password {%s} when call the password validator then it should return an invalid response with TOO_SHORT error', (invalidPassword: string) => {
+        const result = passwordValidator.validate(invalidPassword);
+        expect(result.status).toBeFalsy();
+        expect(result.errors).toEqual([Errors.TOO_SHORT]);
+    });
+
+    it.each([
+        ['More than 15 cha'],
+        ['More than 16 cha'],
+        ['More than 17 cha'],
+    ])('given a long password {%s} when call the password validator then it should return an invalid response with TOO_LONG error', (invalidPassword: string) => {
+        const result = passwordValidator.validate(invalidPassword);
+        expect(result.status).toBeFalsy();
+        expect(result.errors).toEqual([Errors.TOO_LONG]);
+    });
+
+    it.each([
+        ['A perfect p4ass'],
+        ['A p3rfect pass'],
+        ['A perf3ct pass'],
+        ['A p4ss with 15c'],
+        ['P4ssw'],
+    ])('given a valid password {%s} when call the password validator then it should return a valid response with no error', (validPasswords: string) => {
+        const result = passwordValidator.validate(validPasswords);
+        expect(result.status).toBeTruthy();

```

Comment:
Personally, I would not use `toBeTruthy` here. I would really assert that the result is a boolean, it makes my library/API easier to use from a client/user perspective.

For users, it will be easier to just handle 2 values, `true` and `false`. 

Right now you could change your production code to return a `status: null` or `status: undefined` and your tests would still pass, but your users might check for a real boolean `result.status === true`, and that check will fail.